### ISSUE TITLE
[Firefox] Play nicely with hidden tabs

### ISSUE THREAD

**Issue by donbex** [philc/vimium#2998]
Firefox version: 59.0.1
Vimium version: 1.63.2

I'm organising my tabs in different [containers](https://support.mozilla.org/en-US/kb/containers) and I use Conex to keep visible only one at a time — it hides the other tabs through an API introduced in Firefox 59.

The problem is that changing tab with `gt` and `gT` crosses the container boundary. To give an example, suppose that I have two containers, C1 and C2; then I open a tab **t1** at the end of C1, then switch to C2 and open a tab **t2** at the end of C2, then switch back to C1 and open **t3** at then end of C1.

Now, if **t1** is selected and I press `gt` I would expect to select **t3**, while instead Vimium switches to C2 and selects **t2**. Note that switching tab with `ctrl + pgdown` (the default Firefox shortcut) works as expected.

    C1 (visible):    t1      t3
    C2 (hidden):         t2

I'm not sure what you would consider the appropriate behaviour here, but it would be nice to at least have an option to limit tab switching to visible tabs (which I think is the behaviour someone would expect). It would also be nice to have an option/command to switch to the next tab and open a new tab in the current container, even when the others aren't hidden.

**Comment by smblott-github** [comment]
I don't think there's an [API](https://developer.chrome.com/extensions/tabs#type-Tab) for containers.  We'd need that to look at playing nice.

**Comment by smblott-github** [comment]
On further investigation, it looks like this probably isn't to do with containers, but rather [tab hiding](https://wiki.mozilla.org/WebExtensions/TabHiding).  Tab hiding looks to be pretty experimental.  The one thing we would need...

![image](https://user-images.githubusercontent.com/2641335/38021552-12dfd6b4-3275-11e8-9d16-b6bf4b746c7d.png)

... doesn't appear to be implemented.

It looks like we cannot deal with this until they agree on and publish an API.

**Comment by donbex** [comment]
I'm not sure if one can directly use this with `tabs.query()`, but the `tabs.Tabs` type [has the (optional) attribute](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/tabs/Tab) `hidden` for hidden tabs. Would something like this work?

```coffeescript
selectTab = (direction, {count, tab}) ->
  chrome.tabs.query { currentWindow: true, hidden: false }, (tabs) ->
  [...]
```

Or maybe does CoffeeScript offer a way to lazily filter the output of `tabs.query()`?

*P.S.*: I would test it myself, but for some reason I can't get Vimium to compile on my machine. Even after downgrading CoffeScript and NodeJs, I keep getting the error:

    events.js:183
          throw er; // Unhandled 'error' event
          ^
    
    Error: spawn coffee ENOENT
    [...]

**Comment by fturcheti** [comment]
I'm using an extension for grouping and hiding tabs and I'm facing the same issue.

I've tested what @donbex suggested above, not within the Vimium codebase but with a boilerplate extension, and I was able to filter only the visible tabs in Firefox using `hidden: false`. However, Chrome throws an exception because it considers `hidden` to be an unknown property. So maybe something like this could do the trick:

```js
visibleTabs = if Utils.isFirefox() 
    {currentWindow: true, hidden: false} 
  else 
    {currentWindow: true}

chrome.tabs.query(visibleTabs, (tabs) -> [...])
```

**Comment by dakrone** [comment]
I'm also running into this, does anyone know if there is a patch available to Vimium with the suggested fix?

**Comment by UncleSnail** [comment]
> 

In case you are still interested, I made a fork that fixes the simple J/K movement with hidden tabs. I may add other features like moving tabs, and navigating with the numbers in the future, so they all work with hidden tabs, and maybe make a PR someday. But until then, if you want to just have navigation fixed with hidden tabs on Firefox, you can try out my fork. There is a .xpi file in the web-ext-artifacts folder that you can install as an addon if you like.
It works as expected with hidden tabs for me. No promises it will work for you though.
https://github.com/UncleSnail/vimium/tree/master/web-ext-artifacts

**Comment by JeffreyAnimal** [comment]
> > I'm also running into this, does anyone know if there is a patch available to Vimium with the suggested fix?
> 
> In case you are still interested, I made a fork that fixes the simple J/K movement with hidden tabs. I may add other features like moving tabs, and navigating with the numbers in the future, so they all work with hidden tabs, and maybe make a PR someday. But until then, if you want to just have navigation fixed with hidden tabs on Firefox, you can try out my fork. There is a .xpi file in the web-ext-artifacts folder that you can install as an addon if you like. It works as expected with hidden tabs for me. No promises it will work for you though. https://github.com/UncleSnail/vimium/tree/master/web-ext-artifacts

How is the status of the PR? It would be really helpful to have the feature in the upstream Repo. Currently I'm falling back to your Repo regularly since the fix is very useful.

**Comment by UncleSnail** [comment]
> > > I'm also running into this, does anyone know if there is a patch available to Vimium with the suggested fix?
> > 
> > 
> > In case you are still interested, I made a fork that fixes the simple J/K movement with hidden tabs. I may add other features like moving tabs, and navigating with the numbers in the future, so they all work with hidden tabs, and maybe make a PR someday. But until then, if you want to just have navigation fixed with hidden tabs on Firefox, you can try out my fork. There is a .xpi file in the web-ext-artifacts folder that you can install as an addon if you like. It works as expected with hidden tabs for me. No promises it will work for you though. https://github.com/UncleSnail/vimium/tree/master/web-ext-artifacts
> 
> @UncleSnail How is the status of the PR? It would be really helpful to have the feature in the upstream Repo. Currently I'm falling back to your Repo regularly since the fix is very useful.

I've just been using my fork this whole time since I haven't had any problems with it. Every now and then I've thought about updating my solution, I but hadn't decided to yet. For a PR to be reasonable, it should consistently handle hidden tabs across all functionality, which means I would need to fix all the tab navigation for hidden tabs. Now that it looks like no one else is fixing it and there are a few people asking, I'll add this to my list of things to try, and see if I can make a PR. It looks like it would fit in their contribution guidlines, so I think we're good.

**Comment by UncleSnail** [comment]
> @UncleSnail How is the status of the PR? It would be really helpful to have the feature in the upstream Repo. Currently I'm falling back to your Repo regularly since the fix is very useful.

I didn't have anything particularly pressing today, so while it was on my mind, I decided to make a good implementation and submit a PR.
https://github.com/philc/vimium/pull/4477
I implemented hidden tab fixes for all features and I believe I did everything cleanly and well, so we'll see what happens.

**Comment by UncleSnail** [comment]
@JeffreyAnimal If you are interested, I also added a new branch to my fork with a signed xpi file that can be used until something happens with the PR. It should have all the latest Vimium features and full hidden tab support, just like if my changes got merged. That's what I'll be using for now. https://github.com/UncleSnail/vimium/blob/signed-ext/web-ext-artifacts/9af0b42645c4435cbe31-2.1.2.xpi

**Comment by philc** [comment]
This should be fixed by #4477.
